New guide | Adding custom DoH resolvers to DNSCloak (#16)
privacytoolsIO/blog.privacytools.io: collections/_posts/2019-08-29-adding-custom-doh-resolvers-to-dnscloak.md

---
title: Adding Custom DNS over HTTPS Resolvers to DNSCloak
author: nitrohorse
editor: Jonah
layout: post
tags:
  - guides
  - dns
  - dnscrypt
  - ios
---

**[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS, which gives users the ability to encrypt their DNS requests through the use of an on-device VPN profile.

> While highly configurable, its user interface can be unintuitive and doesn't easily allow users to add custom DoH resolvers apart from the default ["public-resolvers" list](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md) that the DNSCrypt project provides.

Reviewer: ```suggestion
While highly configurable, its user interface can be unintuitive to less tech-savvy users and doesn't easily allow users to use custom DoH resolvers, apart from the default ["public-resolvers" list](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md) that the DNSCrypt project provides.
```

> This guide will walk you through setting up DNSCloak to connect to any public resolver that supports either DNSCrypt or DoH.

Reviewer: ```suggestion
In this guide, we'll walk you through setting up DNSCloak on your device, and show you how to connect to any public resolver that supports either DNSCrypt or DoH, even providers not listed in the official DNSCrypt repo, like some of the ones [we recommend on privacytools.io](https://www.privacytools.io/providers/dns/#icanndns).
```

> Before diving in, I'd like to mention that while there is a lot of nuance to DNSCrypt and DoH, it's important to understand that these two DNS protocols essentially achieve the same goals: provide users with the ability to encrypt all DNS traffic to the users' desired [upstream provider(s)](https://www.privacytools.io/providers/dns/#icanndns), while preventing [DNS hijacking](https://en.wikipedia.org/wiki/DNS_hijacking), [spoofing](https://en.wikipedia.org/wiki/DNS_spoofing), and eavesdropping by 3rd parties.

Reviewer: ```suggestion
Before diving in, it's important to understand that while there is a lot of nuance to DNSCrypt and DoH, these two DNS protocols essentially achieve the same goals: They both provide users with the ability to encrypt all DNS traffic to the users' desired [upstream provider(s)](https://www.privacytools.io/providers/dns/#icanndns), while preventing [DNS hijacking](https://en.wikipedia.org/wiki/DNS_hijacking), [spoofing](https://en.wikipedia.org/wiki/DNS_spoofing), and eavesdropping by 3rd parties.
```

> While there is a lot of nuance to DNSCrypt and DoH, it's important to understand that these two DNS protocols essentially achieve the same goals: provide users with the ability to encrypt all DNS traffic to the users' desired [upstream provider(s)](https://www.privacytools.io/providers/dns/#icanndns), while preventing [DNS hijacking](https://en.wikipedia.org/wiki/DNS_hijacking), [spoofing](https://en.wikipedia.org/wiki/DNS_spoofing), and eavesdropping by 3rd parties.

Author: This doesn't seem to fit well here. Nor the paragraph below. Will need to think through this more.

> The development of these DNS protocols is exciting, and unlike Android 9 which has [native support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS unfortunately does not come with any native encrypted DNS support. Thus, DNSCloak fills the gap for iOS users to start benefitting from these protocols now.

Reviewer: Do you think "built-in" sounds better than "native"?

Author: Yeah, it does; good callout!

> And unlike Android 9 which has [native support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS unfortunately does not come with any native support. Thus, iOS users must rely on apps like DNSCloak to benefit from encrypted DNS.

Reviewer: ```suggestion
Additionally, unlike Android 9 which features [native support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS does not come with any native support for encrypted DNS. Instead, iOS users must rely on third-party apps like DNSCloak to benefit from encrypted DNS technologies.
```

Another note I didn't want to include in the suggestion necessarily, but does "built-in" sound better or less technical than "native"?

### Adding a Custom Resolver

> DNSCloak provides a "Config Editor" which allows you to modify various aspects of the program, add custom servers, etc:

Reviewer: > etc

What else does it actually do?

Author: I think it's probably more clear to simply this line to:
> DNSCloak provides a "Config Editor" which allows you to "override or add any [dnscrypt-proxy option](https://github.com/jedisct1/dnscrypt-proxy/wiki/Configuration).":

![](/assets/img/2019-08-24-dnscloak/config-editor.jpeg){: .w-50}

> You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. But, if you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding the [DNS stamp](https://dnscrypt.info/stamps-specifications) of your resolver.

Reviewer: ```suggestion
You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. However, we're only concerned with a single section: If you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding the [DNS stamp](https://dnscrypt.info/stamps-specifications) of your resolver. According to the DNSCrypt project, DNS stamps encode all the parameters required to connect to an encrypted DNS server as a single string. You can think about stamps as QR code, but for DNS.
```

I don't know if this seems too long, but I feel like the explanation of what DNS stamps are shouldn't be on a separate line. Maybe there should be a line break before "However," instead?

Author: Hmm yeah, that makes sense. How about:
> You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. But, if you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding your resolver's [DNS stamp](https://dnscrypt.info/stamps-specifications), an encoded string that contains all the required information needed to connect to an encrypted DNS resolver. You can think about stamps as QR code, but for DNS.

> According to the DNSCrypt project, DNS stamps encode all the parameters required to connect to an encrypted DNS server as a single string. You can think about stamps as QR code, but for DNS.

Reviewer: ```suggestion
```

(see comment/suggestion above/L29)

---
title: Adding Custom DNS over HTTPS Resolvers to DNSCloak
author: nitrohorse
editor: Jonah
layout: post
tags:
  - guides
  - dns
  - dnscrypt
  - ios
---

**[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS, which gives users the ability to encrypt their DNS requests through the use of an on-device VPN profile.

> While highly configurable, its user interface can be unintuitive to less tech-savvy users and doesn't easily allow users to use custom DoH resolvers, apart from the default ["public-resolvers" list](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md) that the DNSCrypt project provides.

Reviewer: ```suggestion
While DNSCloak is highly configurable, its user interface can be unintuitive to less tech-savvy users and doesn't easily allow users to use custom DoH resolvers, apart from the default ["public-resolvers" list](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md) that the DNSCrypt project provides.
```

Before diving in, it's important to understand that while there is a lot of nuance to DNSCrypt and DoH, these two DNS protocols essentially achieve the same goals: They both provide users with the ability to encrypt all DNS traffic to the users' desired [upstream provider(s)](https://www.privacytools.io/providers/dns/#icanndns), while preventing [DNS hijacking](https://en.wikipedia.org/wiki/DNS_hijacking), [spoofing](https://en.wikipedia.org/wiki/DNS_spoofing), and eavesdropping by 3rd parties.

> The development of these DNS protocols is exciting, and unlike Android 9 which has [built-in support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS unfortunately does not come with any built-in encrypted DNS support. Thus, DNSCloak fills the gap for iOS users to start benefitting from these protocols now.

Reviewer: ```suggestion
The development of these DNS protocols is exciting, but unlike Android 9 which has [built-in support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS unfortunately does not come with any built-in encrypted DNS support. Thus, DNSCloak fills the gap for iOS users to start benefitting from these protocols today.
```

> The development of these DNS protocols is exciting, but unlike Android 9 which has [built-in support for DNS over TLS](https://support.google.com/android/answer/9089903) (another protocol with similar goals), iOS unfortunately does not come with any built-in encrypted DNS support. Thus, DNSCloak fills the gap for iOS users to start benefitting from these protocols today.

Reviewer 2: I wonder if you should say yet or if there are any iOS feature request forums that you could link to?

Author: Hmm good call out, it looks like using an [MDM profile could change your DNS](https://stackoverflow.com/questions/19968042/ios-configuration-profiles-to-change-dns-settings) but I’m not sure yet if this is the same thing has [“supervised mode”](https://developer.apple.com/documentation/networkextension/dns_proxy_provider).

Author: >... it appears that the only way to use the new DNS Proxy Network Extension in iOS 11 is through Mobile Device Management (MDM) Tools as a supervised device. This makes it impossible to install for the average user through the App Store.

> An Apple representative confirmed this is currently only for supervised devices but he seems to indicate they are considering opening this up to all devices but with no assurances or timeline.

— https://dnsdisco.com/iOS-dns-proxy-post.html

Author: I could add “yet” and link to this I think.

Author: Updated

This guide will walk you through setting up DNSCloak to connect to any public resolver that supports either DNSCrypt or DoH.

### Adding a Custom Resolver

DNSCloak provides a "Config Editor" which allows you to "override or add any [dnscrypt-proxy option](https://github.com/jedisct1/dnscrypt-proxy/wiki/Configuration)."

![](/assets/img/2019-08-24-dnscloak/config-editor.jpeg){: .w-50}

> You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. But, if you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding your resolver's [DNS stamp](https://dnscrypt.info/stamps-specifications), an encoded string that contains all the required information needed to connect to an encrypted DNS resolver. You can think about stamps as QR code, but for DNS.

Reviewer: > This stamp is for adding your...stamp

Sounds strange to me 🤔 

```suggestion
You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. But, if you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This property is for adding your resolver's [DNS stamp](https://dnscrypt.info/stamps-specifications), an encoded string that contains all the required information needed to connect to an encrypted DNS resolver. You can think about stamps like QR codes, but for DNS.
```

### Generating a Stamp

> Some providers will provide you with a DNS stamp pre-made for you. If your provider does this, great! You can skip ahead to the next section. At the time of writing this post, CZ.NIC's DoH resolver is the only provider [privacytools.io suggests](https://www.privacytools.io/providers/dns/#icanndns) that doesn't provide their users with a DNS stamp on their website, making adoption a bit more difficult. Thankfully however, we can create a DNS stamp ourselves.

Reviewer: ```suggestion
Some providers will provide you with a pre-made DNS stamp for their servers. If your provider does this, great! You can skip ahead to the next section. At the time of writing this post, CZ.NIC's DoH resolver is the only provider [privacytools.io suggests](https://www.privacytools.io/providers/dns/#icanndns) that doesn't provide their users with a DNS stamp on their website, making adoption a bit more difficult. Thankfully however, we can create a DNS stamp ourselves fairly easily.
```

```
[static.'CZ.NIC-193.17.47.1']
stamp = 'sdns://AgMAAAAAAAAACzE5My4xNy40Ny4xAAtvZHZyLm5pYy5jegQvZG9o'
```

![](/assets/img/2019-08-24-dnscloak/config-editor-cz-nic.jpeg){: .w-50}

Select the checkmark icon in the top right corner to save your configuration, and it should be good to go!

![](/assets/img/2019-08-24-dnscloak/dnscloak-cz-nic.jpeg){: .w-50}

Get connected, and we can finally validate DNSCloak is working as expected by visiting [DNSLeakTest.com](https://dnsleaktest.com/):

![](/assets/img/2019-08-24-dnscloak/dnsleaktest-cz-nic.jpeg){: .w-100}

### Adding Cloudflare's Resolver for Firefox

Another public DoH resolver that we may want to use is Cloudflare's public resolver for _Firefox_ [which has a stricter logging policy than Cloudflare's default resolver](https://forum.privacytools.io/t/logging-differences-between-cloudflares-default-dns-over-https-resolver-and-their-resolver-for-firefox/1451).

We can [generate a stamp](https://dnscrypt.info/stamps/) with this information:
- IP address: `1.1.1.1`
- Host name: `mozilla.cloudflare-dns.com`
- Path: `/dns-query`

![](/assets/img/2019-08-24-dnscloak/cloudflare-mozilla-stamp.png){: .w-100}

> You can now paste the following stamp we generated into DNSCloak's Config Editor and start using the resolver.

Reviewer: ```suggestion
You can now paste the stamp we generated into DNSCloak's Config Editor and start using the resolver:
```

```
[static.'Cloudflare Resolver for Firefox']
stamp = 'sdns://AgUAAAAAAAAABzEuMS4xLjEAGm1vemlsbGEuY2xvdWRmbGFyZS1kbnMuY29tCi9kbnMtcXVlcnk'
```

![](/assets/img/2019-08-24-dnscloak/config-editor-cf-moz.jpeg){: .w-50}

![](/assets/img/2019-08-24-dnscloak/dnscloak-cf-moz.jpeg){: .w-50}

### Summary

> Keep in mind that encrypted DNS won't hide the _host name_ (for example, `blog.privacytools.io`) of the sites you visit from your ISP due to [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication#Security_implications)*. If you're looking for anonymity, you should us the [Tor Browser](https://www.torproject.org/) instead. But if you're looking for additional security and increased privacy from 3rd parties, encrypted DNS is a good place to start.

Reviewer 2: ```suggestion
Keep in mind that encrypted DNS won't hide the _host name_ (for example, `blog.privacytools.io`) of the sites you visit from your ISP due to [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication#Security_implications)*. If you're looking for anonymity, you should use the [Tor Browser](https://www.torproject.org/) instead. But if you're looking for additional security and increased privacy from 3rd parties, encrypted DNS is a good place to start.
```
